# Лабораторная работа №3: Аппроксимация функции с помощью двухслойной нейронной сети

**Дисциплина:** Генетические и эволюционные вычисления

**Вариант 3:** $f(t) = (t + 1{,}3)\,\sin(0{,}5t + 1)$, $t \in [-7,\ 7]$, шаг $0{,}1$ (141 точка).

**Цель работы:** изучить архитектуру двухслойной нейронной сети (многослойного персептрона), принципы её обучения методом обратного распространения ошибки и исследовать точность аппроксимации в зависимости от количества нейронов в скрытом слое.

**Постановка (по методическим указаниям):**
- сеть вида `newff`: скрытый слой `tansig`, выходной слой `purelin`;
- обучение `traingd` с параметрами `show = 50`, `lr = 0.05`, `epochs = 600`, `goal = 1e-2`;
- мультипликация: кривая аппроксимации перерисовывается каждые `show = 50` эпох, синхронно строится кривая ошибки;
- серия экспериментов с числом нейронов $S \in \{1, 2, 5, 10, 20, 30\}$.

Сеть и все алгоритмы обучения реализованы с нуля на NumPy и повторяют поведение MATLAB Neural Network Toolbox: нормировка `mapminmax`, инициализация Нгуена — Уидроу (`initnw`), пакетный градиентный спуск `traingd`, критерии останова `epochs` / `goal` / `min_grad`. Для сравнения реализованы `traingdm`, `traingdx` и `trainlm`.

Интерактивное приложение с ползунками, кнопками «Старт / Пауза / Сброс» и живой мультипликацией обучения — `app/nn_trainer.html` (раздел 8).

In [1]:
import time
from dataclasses import dataclass, field
from pathlib import Path

import numpy as np
import pandas as pd
import plotly
import plotly.graph_objects as go
import plotly.io as pio
from plotly.subplots import make_subplots

OUTPUTS = Path('report/outputs')
FIGURES = Path('report/figures')
TABLES = Path('report/tables')
for p in (OUTPUTS, FIGURES, TABLES):
    p.mkdir(parents=True, exist_ok=True)

# Палитра: категориальные цвета проверены на различимость при дальтонизме
PALETTE = ['#2a78d6', '#eb6834', '#1baf7a', '#eda100', '#e87ba4', '#008300', '#4a3aa7', '#e34948']
BLUE_RAMP = ['#b7d3f6', '#86b6ef', '#5598e7', '#2a78d6', '#1c5cab', '#104281', '#0d366b']
INK, INK2, MUTED, GRID, AXIS, SURFACE = '#0b0b0b', '#52514e', '#898781', '#e1e0d9', '#c3c2b7', '#fcfcfb'
GOAL_COLOR = '#d03b3b'

axis_style = dict(gridcolor=GRID, linecolor=AXIS, zeroline=False, ticks='outside', tickcolor=AXIS,
                  ticklen=4, title_font=dict(size=14, color=INK2), tickfont=dict(size=12, color=INK2))
pio.templates['lab'] = go.layout.Template(layout=dict(
    font=dict(family='system-ui, -apple-system, "Segoe UI", Roboto, Arial, sans-serif', size=14, color=INK),
    paper_bgcolor=SURFACE, plot_bgcolor=SURFACE, colorway=PALETTE,
    xaxis=axis_style, yaxis=axis_style,
    title=dict(x=0.01, xanchor='left', font=dict(size=18, color=INK)),
    legend=dict(bgcolor='rgba(252,252,251,0.85)', bordercolor=GRID, borderwidth=1, font=dict(size=12)),
    hoverlabel=dict(bgcolor='white', bordercolor=AXIS, font=dict(size=12, color=INK)),
    margin=dict(l=70, r=30, t=80, b=60),
))
pio.templates.default = 'plotly_white+lab'
PLOT_CONFIG = dict(displaylogo=False, responsive=True,
                   toImageButtonOptions=dict(format='png', scale=2))


def save(fig, name, width=1000, height=520, png=True, png_fig=None):
    """HTML (интерактивный) в report/outputs и PNG (для отчёта) в report/figures."""
    fig.write_html(OUTPUTS / f'{name}.html', include_plotlyjs='cdn', config=PLOT_CONFIG, auto_play=False)
    if png:
        (png_fig or fig).write_image(FIGURES / f'{name}.png', width=width, height=height, scale=2)


def to_latex(df, name, caption, label, col_format=None):
    """Таблица результатов для отчёта: числа в отчёте совпадают с ноутбуком."""
    body = df.to_latex(index=False, escape=False, column_format=col_format or 'c' * len(df.columns))
    body = body.replace('\\toprule', '\\toprule').replace('%', '\\%')
    tex = ('\\begin{table}[H]\n\\caption{' + caption + '}\n\\label{' + label + '}\n\\centering\n\\small\n'
           + body + '\\end{table}\n')
    (TABLES / f'{name}.tex').write_text(tex, encoding='utf-8')


print('numpy', np.__version__, '| pandas', pd.__version__, '| plotly', plotly.__version__)

numpy 2.4.6 | pandas 3.0.6 | plotly 7.1.0


## 1. Исходные данные

Формируем дискретный вектор времени `t = -7:0.1:7` и вектор эталонных значений $y = f(t)$. Для оценки качества **между** узлами обучающей сетки используется плотная контрольная сетка с шагом $0{,}01$ (1401 точка): она не участвует в обучении и позволяет отличить хорошую аппроксимацию от «запоминания» точек.

In [2]:
def f(t):
    return (t + 1.3) * np.sin(0.5 * t + 1)


T0, T1, STEP = -7.0, 7.0, 0.1
t = np.round(np.arange(T0, T1 + STEP / 2, STEP), 10)   # обучающая сетка (как t0:0.1:t1 в MATLAB)
y = f(t)
t_dense = np.linspace(T0, T1, 1401)                     # контрольная сетка
y_dense = f(t_dense)

# экстремумы функции на отрезке
d = np.diff(np.sign(np.diff(y_dense)))
ext_idx = np.where(d != 0)[0] + 1
print(f'Точек обучения: {len(t)}; y ∈ [{y.min():.3f}; {y.max():.3f}]; дисперсия y = {y.var():.3f}')
for i in ext_idx:
    kind = 'максимум' if d[i - 1] < 0 else 'минимум'
    print(f'  локальный {kind}: t = {t_dense[i]:.2f}, f = {y_dense[i]:.3f}')

fig = go.Figure()
fig.add_trace(go.Scatter(x=t_dense, y=y_dense, mode='lines', name='f(t)', line=dict(color=INK2, width=2),
                         hovertemplate='t = %{x:.2f}<br>f(t) = %{y:.3f}<extra></extra>'))
fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='обучающие точки (шаг 0.1)',
                         marker=dict(color=PALETTE[0], size=7, line=dict(color=SURFACE, width=1.5)),
                         hovertemplate='t = %{x:.1f}<br>y = %{y:.3f}<extra></extra>'))
fig.add_trace(go.Scatter(x=t_dense[ext_idx], y=y_dense[ext_idx], mode='markers+text', name='экстремумы',
                         marker=dict(symbol='diamond', size=11, color=PALETTE[1], line=dict(color=SURFACE, width=1.5)),
                         text=[f'{v:.2f}' for v in y_dense[ext_idx]], textposition='top center',
                         textfont=dict(color=INK2, size=12), hoverinfo='skip'))
fig.update_layout(title='Вариант 3: f(t) = (t + 1.3)·sin(0.5t + 1),  t ∈ [−7; 7]',
                  xaxis_title='t', yaxis_title='y', hovermode='closest',
                  legend=dict(x=0.01, y=0.02, yanchor='bottom'))
save(fig, '01_target_function')
fig.show()

Точек обучения: 141; y ∈ [-8.113; 4.274]; дисперсия y = 9.109
  локальный максимум: t = -5.95, f = 4.275
  локальный минимум: t = -1.65, f = -0.061
  локальный максимум: t = 2.18, f = 3.021


## 2. Реализация двухслойной сети (аналог `newff`, `train`, `sim`)

**Архитектура** $1 - S - 1$:
$$a_1 = \mathrm{tansig}(W_1 \tilde t + b_1), \qquad \tilde y = \mathrm{purelin}(W_2 a_1 + b_2) = W_2 a_1 + b_2,$$
где $\tilde t$, $\tilde y$ — вход и выход после нормировки `mapminmax` в $[-1, 1]$ (так делает `newff` по умолчанию). Число настраиваемых параметров $3S + 1$.

**Инициализация Нгуена — Уидроу** (`initnw`): для одного входа $|w_{1,i}| = 0{,}7\,S$, смещения $b_{1,i}$ равномерно расставлены по $[-0{,}7S;\ 0{,}7S]$ — «ступеньки» сигмоид с самого начала распределены по всему отрезку входа, а не собраны в нуле.

**Обучение `traingd`** — пакетный градиентный спуск по ошибке MSE:
$$\theta^{(k+1)} = \theta^{(k)} - \eta\,\nabla_\theta E, \quad E = \frac1N\sum_{i=1}^{N} (\tilde y_i - \tilde y^{*}_i)^2 .$$
Градиенты считаются обратным распространением: $\delta_2 = \tfrac{2}{N}(\tilde y - \tilde y^*)$, $\delta_1 = (\delta_2 W_2) \odot (1 - a_1^2)$.

**Останов** (как в MATLAB): достигнута целевая ошибка `goal` (MSE в исходных единицах $y$), исчерпан лимит `epochs`, норма градиента меньше `min_grad`; дополнительно — расходимость и (при наличии валидационной выборки) `max_fail` ростов ошибки на валидации.

Для сравнения реализованы:
- `traingdm` — градиентный спуск с моментом $mc = 0{,}9$;
- `traingdx` — момент + адаптивный шаг ($\eta \times 1{,}05$ при уменьшении ошибки, $\eta \times 0{,}7$ и откат шага при росте ошибки более чем в $1{,}04$ раза);
- `trainlm` — метод Левенберга — Марквардта: $\Delta\theta = -(J^\top J + \mu I)^{-1} J^\top e$ (алгоритм по умолчанию в `newff`).

In [3]:
class MapMinMax:
    """mapminmax: линейное отображение [min, max] -> [-1, 1] и обратно."""

    def __init__(self, v):
        self.lo, self.hi = float(np.min(v)), float(np.max(v))

    def apply(self, v):
        return 2 * (v - self.lo) / (self.hi - self.lo) - 1

    def reverse(self, v):
        return (v + 1) * (self.hi - self.lo) / 2 + self.lo


class TwoLayerNet:
    """Двухслойная сеть 1-S-1: tansig в скрытом слое, purelin на выходе (аналог newff)."""

    def __init__(self, S, t_train, y_train, seed=0):
        self.S = S
        self.sx, self.sy = MapMinMax(t_train), MapMinMax(y_train)
        rng = np.random.default_rng(seed)
        # Нгуен — Уидроу для одного входа
        beta = 0.7 * S
        self.w1 = np.where(rng.uniform(-1, 1, S) >= 0, 1.0, -1.0) * beta
        self.b1 = (beta * np.linspace(-1, 1, S) * np.sign(self.w1)) if S > 1 else rng.uniform(-beta, beta, 1)
        # выходной слой: малые случайные веса
        self.w2 = rng.uniform(-0.5, 0.5, S)
        self.b2 = float(rng.uniform(-0.5, 0.5))

    @property
    def n_params(self):
        return 3 * self.S + 1

    def get(self):
        return np.concatenate([self.w1, self.b1, self.w2, [self.b2]])

    def set(self, theta):
        S = self.S
        self.w1, self.b1, self.w2 = theta[:S].copy(), theta[S:2 * S].copy(), theta[2 * S:3 * S].copy()
        self.b2 = float(theta[3 * S])

    def forward(self, xn):
        a1 = np.tanh(np.outer(xn, self.w1) + self.b1)      # tansig
        return a1, a1 @ self.w2 + self.b2                   # purelin

    def gradient(self, xn, err):
        """Обратное распространение: градиент MSE (в нормированном пространстве)."""
        a1, _ = self.forward(xn)
        delta2 = 2.0 * err / len(xn)
        delta1 = np.outer(delta2, self.w2) * (1.0 - a1 ** 2)
        return np.concatenate([delta1.T @ xn, delta1.sum(0), a1.T @ delta2, [delta2.sum()]])

    def jacobian(self, xn):
        """Якобиан выходов сети по параметрам (для Левенберга — Марквардта)."""
        a1, _ = self.forward(xn)
        d1 = (1.0 - a1 ** 2) * self.w2
        return np.hstack([d1 * xn[:, None], d1, a1, np.ones((len(xn), 1))])

    def sim(self, t_in):
        """Аналог sim(net, t): выход в исходных единицах."""
        return self.sy.reverse(self.forward(self.sx.apply(t_in))[1])

    def hidden_contributions(self, t_in):
        """Вклад каждого нейрона w2_i * tansig(...) в выход (в исходных единицах, без смещения)."""
        a1, _ = self.forward(self.sx.apply(t_in))
        return a1 * self.w2 * (self.sy.hi - self.sy.lo) / 2


@dataclass
class TrainRecord:
    fcn: str
    perf: list = field(default_factory=list)       # MSE на обучающей выборке по эпохам (исходные единицы)
    vperf: list = field(default_factory=list)      # MSE на валидации (если задана)
    grad: list = field(default_factory=list)
    lr: list = field(default_factory=list)
    snapshots: list = field(default_factory=list)  # (эпоха, выход сети на контрольной сетке)
    stop: str = ''
    epochs: int = 0
    seconds: float = 0.0

    @property
    def final(self):
        return self.perf[-1]


def train(net, t_tr, y_tr, fcn='traingd', lr=0.05, epochs=600, goal=1e-2, show=50, min_grad=None,
          mc=0.9, lr_inc=1.05, lr_dec=0.7, max_perf_inc=1.04,
          mu=1e-3, mu_dec=0.1, mu_inc=10.0, mu_max=1e10,
          val=None, max_fail=6, snap_t=None, verbose=False):
    """Аналог train(net, t, y) с net.trainFcn = fcn и параметрами net.trainParam.*"""
    if min_grad is None:
        min_grad = 1e-7 if fcn == 'trainlm' else 1e-5
    xn, yn = net.sx.apply(t_tr), net.sy.apply(y_tr)
    rec = TrainRecord(fcn)
    tic = time.perf_counter()

    def perf_orig(xn_, y_):
        return float(np.mean((net.sy.reverse(net.forward(xn_)[1]) - y_) ** 2))

    if val is not None:
        xv, yv = net.sx.apply(val[0]), val[1]
        best_v, best_theta, fails = np.inf, net.get(), 0

    theta = net.get()
    dtheta = np.zeros_like(theta)
    ep = 0
    while True:
        err = net.forward(xn)[1] - yn
        p = perf_orig(xn, y_tr)
        g = net.gradient(xn, err)
        gnorm = float(np.linalg.norm(g))
        rec.perf.append(p); rec.grad.append(gnorm); rec.lr.append(lr)
        if val is not None:
            pv = perf_orig(xv, yv)
            rec.vperf.append(pv)
            if pv < best_v:
                best_v, best_theta, fails = pv, net.get(), 0
            elif ep > 0:
                fails += 1

        if snap_t is not None and ep % show == 0:
            rec.snapshots.append((ep, net.sim(snap_t)))
        if verbose and ep % show == 0:
            print(f'{fcn.upper()}, Epoch {ep}/{epochs}, MSE {p:.5g}/{goal:g}, Gradient {gnorm:.4g}/{min_grad:g}')

        stop = None
        if not np.isfinite(p) or p > 1e8:
            stop = 'divergence'
        elif p <= goal:
            stop = 'goal'
        elif ep >= epochs:
            stop = 'epochs'
        elif gnorm < min_grad:
            stop = 'min_grad'
        elif val is not None and fails >= max_fail:
            stop = 'val_fail'
        if stop:
            break

        # --- один шаг обучения ---
        if fcn == 'traingd':
            theta = theta - lr * g
        elif fcn == 'traingdm':
            dtheta = mc * dtheta - (1 - mc) * lr * g
            theta = theta + dtheta
        elif fcn == 'traingdx':
            d_new = mc * dtheta - lr * g
            net.set(theta + d_new)
            p_new = perf_orig(xn, y_tr)
            if p_new > p * max_perf_inc:            # шаг неудачен: откат, уменьшаем lr
                lr *= lr_dec
                dtheta = np.zeros_like(theta)
            else:
                if p_new < p:
                    lr *= lr_inc
                theta, dtheta = theta + d_new, d_new
        elif fcn == 'trainlm':
            J = net.jacobian(xn)
            JtJ, Jte = J.T @ J, J.T @ err
            e0 = float(np.mean(err ** 2))
            while True:
                step = np.linalg.solve(JtJ + mu * np.eye(len(theta)), Jte)
                net.set(theta - step)
                if float(np.mean((net.forward(xn)[1] - yn) ** 2)) < e0:
                    theta = theta - step
                    mu = max(mu * mu_dec, 1e-20)
                    break
                mu *= mu_inc
                if mu > mu_max:
                    break
            if mu > mu_max:
                net.set(theta)
                rec.stop = 'mu_max'
                break
        else:
            raise ValueError(fcn)
        net.set(theta)
        ep += 1

    if not rec.stop:
        rec.stop = stop
    if val is not None:
        net.set(best_theta)          # как в MATLAB: веса с минимальной ошибкой на валидации
    if snap_t is not None and (not rec.snapshots or rec.snapshots[-1][0] != ep):
        rec.snapshots.append((ep, net.sim(snap_t)))
    if verbose:
        print(f'{fcn.upper()}, Epoch {ep}/{epochs}, MSE {rec.perf[-1]:.5g}/{goal:g}, '
              f'Gradient {rec.grad[-1]:.4g}/{min_grad:g}  -> останов: {rec.stop}')
    rec.epochs = ep
    rec.seconds = time.perf_counter() - tic
    return rec


def mse(a, b):
    return float(np.mean((a - b) ** 2))


def r2(pred, true):
    return 1 - np.sum((pred - true) ** 2) / np.sum((true - true.mean()) ** 2)

**Проверка корректности обратного распространения.** Сравним аналитический градиент с численным (центральные разности). Относительная ошибка порядка $10^{-9}$ и меньше означает, что формулы backpropagation реализованы верно.

In [4]:
net_chk = TwoLayerNet(7, t, y, seed=123)
xn_chk, yn_chk = net_chk.sx.apply(t), net_chk.sy.apply(y)
theta0 = net_chk.get()
g_an = net_chk.gradient(xn_chk, net_chk.forward(xn_chk)[1] - yn_chk)
g_num = np.zeros_like(theta0)
for i in range(len(theta0)):
    for sgn in (+1, -1):
        th = theta0.copy(); th[i] += sgn * 1e-6; net_chk.set(th)
        g_num[i] += sgn * np.mean((net_chk.forward(xn_chk)[1] - yn_chk) ** 2) / 2e-6
net_chk.set(theta0)
rel = np.linalg.norm(g_an - g_num) / np.linalg.norm(g_an + g_num)
print(f'Относительная ошибка градиента: {rel:.2e}')

J_num = np.zeros((len(t), len(theta0)))
for i in range(len(theta0)):
    th = theta0.copy(); th[i] += 1e-6; net_chk.set(th); up = net_chk.forward(xn_chk)[1]
    th[i] -= 2e-6; net_chk.set(th); dn = net_chk.forward(xn_chk)[1]
    J_num[:, i] = (up - dn) / 2e-6
net_chk.set(theta0)
print(f'Относительная ошибка якобиана: {np.linalg.norm(net_chk.jacobian(xn_chk) - J_num) / np.linalg.norm(J_num):.2e}')

Относительная ошибка градиента: 4.10e-11
Относительная ошибка якобиана: 1.03e-10


## 3. Обучение по заданию с мультипликацией (S = 10)

Параметры строго по методическим указаниям: `traingd`, `lr = 0.05`, `epochs = 600`, `goal = 1e-2`, `show = 50`. Каждые 50 эпох печатается строка протокола (как в окне MATLAB) и сохраняется снимок кривой аппроксимации.

In [5]:
S_MAIN, SEED_MAIN = 10, 0
net_main = TwoLayerNet(S_MAIN, t, y, seed=SEED_MAIN)
rec_main = train(net_main, t, y, fcn='traingd', lr=0.05, epochs=600, goal=1e-2, show=50,
                 snap_t=t_dense, verbose=True)
y_main = net_main.sim(t_dense)
print(f'\nИтог: MSE(обуч.) = {rec_main.final:.4f}; MSE(контроль) = {mse(y_main, y_dense):.4f}; '
      f'R² = {r2(y_main, y_dense):.5f}; max|ошибка| = {np.max(np.abs(y_main - y_dense)):.3f}')

TRAINGD, Epoch 0/600, MSE 144.86/0.01, Gradient 8.435/1e-05
TRAINGD, Epoch 50/600, MSE 0.79958/0.01, Gradient 0.0934/1e-05
TRAINGD, Epoch 100/600, MSE 0.46114/0.01, Gradient 0.03874/1e-05
TRAINGD, Epoch 150/600, MSE 0.35067/0.01, Gradient 0.03055/1e-05
TRAINGD, Epoch 200/600, MSE 0.27303/0.01, Gradient 0.02646/1e-05
TRAINGD, Epoch 250/600, MSE 0.21409/0.01, Gradient 0.02315/1e-05
TRAINGD, Epoch 300/600, MSE 0.16877/0.01, Gradient 0.02034/1e-05
TRAINGD, Epoch 350/600, MSE 0.13367/0.01, Gradient 0.01793/1e-05
TRAINGD, Epoch 400/600, MSE 0.10635/0.01, Gradient 0.01584/1e-05
TRAINGD, Epoch 450/600, MSE 0.085/0.01, Gradient 0.01401/1e-05
TRAINGD, Epoch 500/600, MSE 0.068251/0.01, Gradient 0.01242/1e-05
TRAINGD, Epoch 550/600, MSE 0.05508/0.01, Gradient 0.01102/1e-05
TRAINGD, Epoch 600/600, MSE 0.044695/0.01, Gradient 0.009792/1e-05
TRAINGD, Epoch 600/600, MSE 0.044695/0.01, Gradient 0.009792/1e-05  -> останов: epochs

Итог: MSE(обуч.) = 0.0447; MSE(контроль) = 0.0443; R² = 0.99501; max|ошиб

In [6]:
# Мультипликация: кривая аппроксимации каждые show = 50 эпох + синхронная кривая ошибки
snaps = rec_main.snapshots
perf_main = np.array(rec_main.perf)
ep_axis = np.arange(len(perf_main))

fig = make_subplots(rows=1, cols=2, column_widths=[0.58, 0.42], horizontal_spacing=0.09,
                    subplot_titles=('Аппроксимация', 'Ошибка обучения (MSE)'))
fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='f(t), обучающие точки',
                         marker=dict(color=INK2, size=5, opacity=0.75),
                         hovertemplate='t = %{x:.1f}<br>f = %{y:.3f}<extra></extra>'), 1, 1)
fig.add_trace(go.Scatter(x=t_dense, y=snaps[0][1], mode='lines', name='выход сети',
                         line=dict(color=PALETTE[1], width=3),
                         hovertemplate='t = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'), 1, 1)
fig.add_trace(go.Scatter(x=ep_axis[:1], y=perf_main[:1], mode='lines', name='MSE',
                         line=dict(color=PALETTE[0], width=2),
                         hovertemplate='эпоха %{x}<br>MSE = %{y:.4g}<extra></extra>'), 1, 2)
fig.add_trace(go.Scatter(x=[0, 600], y=[1e-2, 1e-2], mode='lines', name='goal = 1e-2',
                         line=dict(color=GOAL_COLOR, width=1.5, dash='dash'), hoverinfo='skip'), 1, 2)
fig.add_trace(go.Scatter(x=[0], y=[perf_main[0]], mode='markers', showlegend=False,
                         marker=dict(color=PALETTE[0], size=9, line=dict(color=SURFACE, width=2)),
                         hoverinfo='skip'), 1, 2)

frames = []
for ep, yhat in snaps:
    frames.append(go.Frame(name=str(ep), data=[
        go.Scatter(x=t_dense, y=yhat),
        go.Scatter(x=ep_axis[:ep + 1], y=perf_main[:ep + 1]),
        go.Scatter(x=[ep], y=[perf_main[ep]]),
    ], traces=[1, 2, 4], layout=go.Layout(annotations=list(fig.layout.annotations) + [dict(
        text=f'<b>Эпоха {ep}</b> · MSE = {perf_main[ep]:.4f}', xref='paper', yref='paper', x=0.0, y=1.17,
        showarrow=False, font=dict(size=14, color=INK), xanchor='left')])))
fig.frames = frames
ylo, yhi = y.min() - 1.5, y.max() + 1.5
fig.update_yaxes(range=[ylo, yhi], title_text='y', row=1, col=1)
fig.update_xaxes(title_text='t', row=1, col=1)
fig.update_yaxes(type='log', range=[np.log10(5e-3), np.log10(perf_main.max() * 1.5)], title_text='MSE', row=1, col=2)
fig.update_xaxes(range=[0, 600], title_text='эпоха', row=1, col=2)
play = dict(frame=dict(duration=700, redraw=False), transition=dict(duration=550, easing='cubic-in-out'),
            fromcurrent=True, mode='immediate')
fig.update_layout(
    title=f'Мультипликация обучения traingd: S = {S_MAIN}, lr = 0.05, show = 50',
    height=580, margin=dict(t=120, b=110),
    legend=dict(x=0.995, xanchor='right', y=0.985, yanchor='top'),
    annotations=list(fig.layout.annotations) + [dict(
        text=f'<b>Эпоха 0</b> · MSE = {perf_main[0]:.4f}', xref='paper', yref='paper', x=0.0, y=1.17,
        showarrow=False, font=dict(size=14, color=INK), xanchor='left')],
    updatemenus=[dict(type='buttons', direction='left', x=1.0, xanchor='right', y=1.2, yanchor='bottom',
                      showactive=False, pad=dict(r=0, t=0), buttons=[
                          dict(label='▶ Старт', method='animate', args=[None, play]),
                          dict(label='⏸ Пауза', method='animate',
                               args=[[None], dict(frame=dict(duration=0, redraw=False), mode='immediate')]),
                          dict(label='⟲ Сброс', method='animate',
                               args=[['0'], dict(frame=dict(duration=0, redraw=False), mode='immediate',
                                                 transition=dict(duration=0))]),
                      ])],
    sliders=[dict(active=0, x=0.0, len=1.0, y=-0.05, yanchor='top', pad=dict(t=40),
                  currentvalue=dict(prefix='эпоха: ', font=dict(size=13, color=INK2)),
                  steps=[dict(label=str(ep), method='animate',
                              args=[[str(ep)], dict(frame=dict(duration=0, redraw=False), mode='immediate',
                                                    transition=dict(duration=300))]) for ep, _ in snaps])],
)
# PNG — финальное состояние
png = go.Figure(fig)
png.frames = []
png.data[1].y = snaps[-1][1]
png.data[2].x, png.data[2].y = ep_axis, perf_main
png.data[4].x, png.data[4].y = [ep_axis[-1]], [perf_main[-1]]
png.layout.updatemenus = ()
png.layout.sliders = ()
png.update_layout(margin=dict(b=60),
                  annotations=list(fig.layout.annotations[:2]) + [dict(
                      text=f'<b>Эпоха {rec_main.epochs}</b> · MSE = {perf_main[-1]:.4f}', xref='paper',
                      yref='paper', x=0, y=1.17, showarrow=False, font=dict(size=14), xanchor='left')])
save(fig, '02_training_animation', width=1200, height=520, png_fig=png)
fig.show()

In [7]:
# Раскадровка «подстройки» кривой (для отчёта): снимки через каждые 100 эпох
fig = go.Figure()
fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='f(t)', marker=dict(color=INK2, size=5, opacity=0.6)))
shown = [s for s in snaps if s[0] in (0, 50, 100, 200, 300, 400, 600)]
for (ep, yhat), col in zip(shown, BLUE_RAMP):
    fig.add_trace(go.Scatter(x=t_dense, y=yhat, mode='lines', name=f'эпоха {ep}',
                             line=dict(color=col, width=2.5 if ep == shown[-1][0] else 2),
                             hovertemplate=f'эпоха {ep}<br>' + 't = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'))
fig.update_layout(title='Изгибание кривой аппроксимации в ходе обучения (traingd, S = 10)',
                  xaxis_title='t', yaxis_title='y', legend=dict(x=0.01, y=0.02, yanchor='bottom'))
save(fig, '03_training_snapshots')
fig.show()

## 4. Серия экспериментов: S = 1, 2, 5, 10, 20, 30 (параметры по заданию)

Результат градиентного спуска зависит от начальных весов, поэтому для каждого $S$ выполнено **10 запусков с разной инициализацией** и приводятся медиана, лучший и худший результат. Для графиков выбирается запуск с медианной ошибкой — «типичная» сеть данного размера, а не удачный выброс.

In [8]:
S_LIST = [1, 2, 5, 10, 20, 30]
SEEDS = range(10)


def run_series(fcn, S_list=S_LIST, seeds=SEEDS, **kw):
    out = {}
    for S in S_list:
        runs = []
        for s in seeds:
            net = TwoLayerNet(S, t, y, seed=s)
            rec = train(net, t, y, fcn=fcn, **kw)
            yh = net.sim(t_dense)
            runs.append(dict(seed=s, rec=rec, net=net, y_dense=yh, test=mse(yh, y_dense)))
        out[S] = runs
    return out


def median_run(runs):
    order = np.argsort([r['rec'].final for r in runs])
    return runs[order[len(order) // 2]]


def summarize(series, name):
    rows = []
    for S, runs in series.items():
        fin = np.array([r['rec'].final for r in runs])
        tst = np.array([r['test'] for r in runs])
        eps = np.array([r['rec'].epochs for r in runs])
        ok = np.array([r['rec'].stop == 'goal' for r in runs])
        m = median_run(runs)
        rows.append(dict(алгоритм=name, S=S, параметров=3 * S + 1,
                         **{'MSE медиана': np.median(fin), 'MSE лучшая': fin.min(), 'MSE худшая': fin.max(),
                            'MSE контроль': np.median(tst), 'R² контроль': r2(m['y_dense'], y_dense),
                            'эпох (медиана)': int(np.median(eps)), 'цель достигнута': f'{ok.sum()}/{len(ok)}',
                            'время, мс': 1000 * np.median([r['rec'].seconds for r in runs])}))
    return pd.DataFrame(rows)


tic = time.perf_counter()
series_gd = run_series('traingd', lr=0.05, epochs=600, goal=1e-2, show=50)
df_gd = summarize(series_gd, 'traingd')
print(f'Серия выполнена за {time.perf_counter() - tic:.1f} с')
df_gd.style.format({'MSE медиана': '{:.4f}', 'MSE лучшая': '{:.4f}', 'MSE худшая': '{:.4f}',
                    'MSE контроль': '{:.4f}', 'R² контроль': '{:.5f}', 'время, мс': '{:.0f}'}).hide(axis='index')

Серия выполнена за 1.6 с


алгоритм,S,параметров,MSE медиана,MSE лучшая,MSE худшая,MSE контроль,R² контроль,эпох (медиана),цель достигнута,"время, мс"
traingd,1,4,4.0830,3.2209,4.8133,4.0356,0.53346,600,0/10,11
traingd,2,7,2.1762,2.1025,2.2680,2.1764,0.75350,600,0/10,14
traingd,5,16,0.0484,0.0280,0.1664,0.0436,0.99478,600,0/10,17
traingd,10,31,0.0493,0.0199,0.0840,0.0471,0.99440,600,0/10,29
traingd,20,61,0.0670,0.0272,0.1045,0.0636,0.99230,600,0/10,39
traingd,30,91,0.0958,0.0341,0.1208,0.0955,0.98905,600,0/10,52


In [9]:
# Аппроксимации (типичный запуск) — малые множители 2×3
fig = make_subplots(rows=2, cols=3, shared_yaxes=True, horizontal_spacing=0.04, vertical_spacing=0.16,
                    subplot_titles=[f'S = {S}: MSE = {median_run(series_gd[S])["rec"].final:.3f}' for S in S_LIST])
for k, S in enumerate(S_LIST):
    r, c = k // 3 + 1, k % 3 + 1
    m = median_run(series_gd[S])
    fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='f(t)', legendgroup='f', showlegend=k == 0,
                             marker=dict(color=INK2, size=4, opacity=0.6),
                             hovertemplate='t = %{x:.1f}<br>f = %{y:.3f}<extra></extra>'), r, c)
    fig.add_trace(go.Scatter(x=t_dense, y=m['y_dense'], mode='lines', name='выход сети', legendgroup='nn',
                             showlegend=k == 0, line=dict(color=PALETTE[1], width=2.5),
                             hovertemplate=f'S = {S}<br>' + 't = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'), r, c)
    fig.update_xaxes(title_text='t' if r == 2 else None, row=r, col=c)
fig.update_yaxes(range=[ylo, yhi])
fig.update_layout(title='traingd, lr = 0.05, 600 эпох: аппроксимация при разном числе нейронов',
                  height=640, legend=dict(orientation='h', x=0.5, xanchor='center', y=-0.1))
save(fig, '04_series_approximations', width=1200, height=640)
fig.show()

In [10]:
# Кривые обучения (типичный запуск каждого S)
fig = go.Figure()
for k, S in enumerate(S_LIST):
    pr = median_run(series_gd[S])['rec'].perf
    fig.add_trace(go.Scatter(x=np.arange(len(pr)), y=pr, mode='lines', name=f'S = {S}',
                             line=dict(color=PALETTE[k], width=2),
                             hovertemplate=f'S = {S}<br>' + 'эпоха %{x}<br>MSE = %{y:.4g}<extra></extra>'))
fig.add_hline(y=1e-2, line=dict(color=GOAL_COLOR, dash='dash', width=1.5),
              annotation_text='goal = 1e-2', annotation_position='bottom right', annotation_font_color=GOAL_COLOR)
fig.update_layout(title='Падение ошибки traingd при разном S (lr = 0.05)', xaxis_title='эпоха',
                  yaxis_title='MSE (log)', yaxis_type='log', hovermode='x unified')
save(fig, '05_series_performance')
fig.show()

In [11]:
# Разброс итоговой ошибки по 10 инициализациям
fig = go.Figure()
for k, S in enumerate(S_LIST):
    fin = [r['rec'].final for r in series_gd[S]]
    fig.add_trace(go.Box(y=fin, x=[f'S = {S}'] * len(fin), name=f'S = {S}', boxpoints='all', jitter=0.4,
                         pointpos=0, marker=dict(color=PALETTE[0], size=8, line=dict(color=SURFACE, width=1.5)),
                         line=dict(color=INK2, width=1.5), fillcolor='rgba(42,120,214,0.10)', showlegend=False,
                         hovertemplate='%{x}<br>MSE = %{y:.4f}<extra></extra>'))
fig.add_hline(y=1e-2, line=dict(color=GOAL_COLOR, dash='dash', width=1.5),
              annotation_text='goal = 1e-2', annotation_position='bottom right', annotation_font_color=GOAL_COLOR)
fig.update_layout(title='Итоговая MSE после 600 эпох traingd: 10 инициализаций на каждое S',
                  yaxis_type='log', yaxis_title='MSE (log)', xaxis_title=None)
save(fig, '06_series_mse_spread')
fig.show()

## 5. Доведение обучения до целевой ошибки: какие алгоритмы действительно обучают сеть

600 эпох `traingd` с $\eta = 0{,}05$ — слишком короткий бюджет. Проверим, что сеть **способна** достичь `goal = 1e-2`, если дать ей нормальное обучение:
- `traingd` без ограничения в 600 эпох (до 20 000);
- `traingdx` — градиентный спуск с моментом и адаптивным шагом, те же 600 эпох;
- `trainlm` — Левенберг — Марквардт, те же 600 эпох.

In [12]:
tic = time.perf_counter()
series_gd_long = run_series('traingd', lr=0.05, epochs=20000, goal=1e-2)
series_gdx = run_series('traingdx', lr=0.05, epochs=600, goal=1e-2)
series_lm = run_series('trainlm', epochs=600, goal=1e-2)
print(f'Готово за {time.perf_counter() - tic:.1f} с')
df_algos = pd.concat([df_gd, summarize(series_gd_long, 'traingd (≤20000 эп.)'),
                      summarize(series_gdx, 'traingdx'), summarize(series_lm, 'trainlm')], ignore_index=True)
df_algos[['алгоритм', 'S', 'MSE медиана', 'MSE контроль', 'эпох (медиана)', 'цель достигнута', 'время, мс']] \
    .style.format({'MSE медиана': '{:.4f}', 'MSE контроль': '{:.4f}', 'время, мс': '{:.0f}'}).hide(axis='index')

Готово за 17.7 с


алгоритм,S,MSE медиана,MSE контроль,эпох (медиана),цель достигнута,"время, мс"
traingd,1,4.0830,4.0356,600,0/10,11
traingd,2,2.1762,2.1764,600,0/10,14
traingd,5,0.0484,0.0436,600,0/10,17
traingd,10,0.0493,0.0471,600,0/10,29
traingd,20,0.0670,0.0636,600,0/10,39
traingd,30,0.0958,0.0955,600,0/10,52
traingd (≤20000 эп.),1,1.5527,1.5578,20000,0/10,379
traingd (≤20000 эп.),2,0.8646,0.8664,20000,0/10,506
traingd (≤20000 эп.),5,0.0100,0.0090,3673,10/10,111
traingd (≤20000 эп.),10,0.0100,0.0090,2360,10/10,114


In [13]:
# Сравнение скорости сходимости алгоритмов при S = 10 (типичный запуск)
algo_series = [('traingd, lr = 0.05', series_gd_long), ('traingdx', series_gdx), ('trainlm', series_lm)]
fig = go.Figure()
for k, (name, ser) in enumerate(algo_series):
    pr = median_run(ser[10])['rec'].perf
    fig.add_trace(go.Scatter(x=np.arange(1, len(pr) + 1), y=pr, mode='lines+markers' if len(pr) < 30 else 'lines',
                             name=f'{name}: {len(pr) - 1} эп.', line=dict(color=PALETTE[k], width=2),
                             marker=dict(size=8, line=dict(color=SURFACE, width=1.5)),
                             hovertemplate=f'{name}<br>' + 'эпоха %{x}<br>MSE = %{y:.4g}<extra></extra>'))
fig.add_hline(y=1e-2, line=dict(color=GOAL_COLOR, dash='dash', width=1.5),
              annotation_text='goal = 1e-2', annotation_position='top right', annotation_font_color=GOAL_COLOR)
fig.update_layout(title='S = 10: сколько эпох нужно разным алгоритмам, чтобы достичь goal',
                  xaxis_title='эпоха (log)', yaxis_title='MSE (log)', xaxis_type='log', yaxis_type='log',
                  legend=dict(x=0.99, xanchor='right', y=0.99))
save(fig, '07_algorithms_convergence')
fig.show()

### Предельная точность при заданном числе нейронов

Чтобы отделить **ёмкость сети** от **качества оптимизации**, обучим сети $S = 1 \dots 30$ методом Левенберга — Марквардта практически до сходимости (`goal = 1e-10`, до 500 эпох, лучший из 10 запусков). Получается нижняя граница ошибки, которую в принципе может дать сеть с $S$ нейронами на этой функции.

In [14]:
S_ALL = list(range(1, 31))
cap_rows, cap_models = [], {}
tic = time.perf_counter()
for S in S_ALL:
    best = None
    for s in SEEDS:
        net = TwoLayerNet(S, t, y, seed=s)
        rec = train(net, t, y, fcn='trainlm', epochs=500, goal=1e-10)
        if best is None or rec.final < best[1].final:
            best = (net, rec)
    net, rec = best
    yh = net.sim(t_dense)
    cap_models[S] = net
    cap_rows.append(dict(S=S, train=rec.final, test=mse(yh, y_dense), maxerr=float(np.max(np.abs(yh - y_dense)))))
df_cap = pd.DataFrame(cap_rows)
print(f'Готово за {time.perf_counter() - tic:.1f} с')
S_min_goal = int(df_cap.loc[df_cap.train <= 1e-2, 'S'].min())
print(f'Минимальное S, при котором достижима goal = 1e-2: S = {S_min_goal}')
df_cap.head(8).style.format({'train': '{:.2e}', 'test': '{:.2e}', 'maxerr': '{:.2e}'}).hide(axis='index')

Готово за 17.2 с
Минимальное S, при котором достижима goal = 1e-2: S = 3


S,train,test,maxerr
1,1.50e+00,1.51e+00,2.17e+00
2,6.75e-01,6.77e-01,1.62e+00
3,9.25e-05,8.46e-05,4.39e-02
4,8.71e-06,8.32e-06,7.81e-03
5,7.27e-09,6.90e-09,2.92e-04
6,9.91e-11,9.67e-11,2.95e-05
7,9.72e-11,9.51e-11,2.43e-05
8,1.26e-08,1.18e-08,3.70e-04


In [15]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=df_cap.S, y=df_cap.train, mode='lines+markers', name='обучающая сетка (шаг 0.1)',
                         line=dict(color=PALETTE[0], width=2), marker=dict(size=8, line=dict(color=SURFACE, width=1.5)),
                         hovertemplate='S = %{x}<br>MSE обуч. = %{y:.2e}<extra></extra>'))
fig.add_trace(go.Scatter(x=df_cap.S, y=df_cap.test, mode='lines+markers', name='контрольная сетка (шаг 0.01)',
                         line=dict(color=PALETTE[1], width=2, dash='dot'),
                         marker=dict(size=8, symbol='diamond', line=dict(color=SURFACE, width=1.5)),
                         hovertemplate='S = %{x}<br>MSE контроль = %{y:.2e}<extra></extra>'))
gd_med = df_gd.set_index('S')['MSE медиана']
fig.add_trace(go.Scatter(x=gd_med.index, y=gd_med.values, mode='markers', name='traingd по заданию (медиана)',
                         marker=dict(size=11, symbol='x', color=PALETTE[2], line=dict(width=0)),
                         hovertemplate='S = %{x}<br>traingd 600 эп.: %{y:.4f}<extra></extra>'))
fig.add_hline(y=1e-2, line=dict(color=GOAL_COLOR, dash='dash', width=1.5),
              annotation_text='goal = 1e-2', annotation_position='top right', annotation_font_color=GOAL_COLOR)
fig.update_layout(title='Точность аппроксимации в зависимости от числа нейронов скрытого слоя',
                  xaxis_title='число нейронов S', yaxis_title='MSE (log)', yaxis_type='log',
                  xaxis=dict(dtick=2), hovermode='x unified', legend=dict(x=0.99, xanchor='right', y=0.99))
save(fig, '08_capacity_curve')
fig.show()

In [16]:
# Хорошо обученные сети (trainlm до сходимости): видно, где кончается ёмкость и начинается точная аппроксимация
S_SHOW = [1, 2, 3, 5, 10, 30]
cap = df_cap.set_index('S')
fig = make_subplots(rows=2, cols=3, shared_yaxes=True, horizontal_spacing=0.04, vertical_spacing=0.16,
                    subplot_titles=[f'S = {S}: MSE = {cap.train[S]:.1e}' for S in S_SHOW])
for k, S in enumerate(S_SHOW):
    r, c = k // 3 + 1, k % 3 + 1
    fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='f(t)', legendgroup='f', showlegend=k == 0,
                             marker=dict(color=INK2, size=4, opacity=0.6),
                             hovertemplate='t = %{x:.1f}<br>f = %{y:.3f}<extra></extra>'), r, c)
    fig.add_trace(go.Scatter(x=t_dense, y=cap_models[S].sim(t_dense), mode='lines', name='выход сети (trainlm)',
                             legendgroup='nn', showlegend=k == 0, line=dict(color=PALETTE[0], width=2.5),
                             hovertemplate=f'S = {S}<br>' + 't = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'), r, c)
    fig.update_xaxes(title_text='t' if r == 2 else None, row=r, col=c)
fig.update_yaxes(range=[ylo, yhi])
fig.update_layout(title='Сети, обученные до сходимости (trainlm, лучший из 10 запусков)',
                  height=640, legend=dict(orientation='h', x=0.5, xanchor='center', y=-0.1))
save(fig, '09_well_trained_models', width=1200, height=640)
fig.show()

In [17]:
# Как сеть складывает функцию из сигмоид: вклады нейронов (S = 3 и S = 5, trainlm)
fig = make_subplots(rows=1, cols=2, shared_yaxes=True, horizontal_spacing=0.05,
                    subplot_titles=[f'S = {S}: MSE = {df_cap.set_index("S").train[S]:.1e}' for S in (3, 5)])
for c, S in enumerate((3, 5), start=1):
    net = cap_models[S]
    contrib = net.hidden_contributions(t_dense)
    for i in range(S):
        fig.add_trace(go.Scatter(x=t_dense, y=contrib[:, i], mode='lines', name=f'нейрон {i + 1}',
                                 legendgroup=f'n{i}', showlegend=c == 2, line=dict(color=PALETTE[i + 1], width=1.8),
                                 hovertemplate=f'нейрон {i + 1}<br>' + 't = %{x:.2f}<br>вклад = %{y:.3f}<extra></extra>'),
                      1, c)
    fig.add_trace(go.Scatter(x=t, y=y, mode='markers', name='f(t)', legendgroup='f', showlegend=c == 2,
                             marker=dict(color=INK2, size=4, opacity=0.5)), 1, c)
    fig.add_trace(go.Scatter(x=t_dense, y=net.sim(t_dense), mode='lines', name='выход сети = Σ вкладов + b',
                             legendgroup='out', showlegend=c == 2, line=dict(color=PALETTE[0], width=3),
                             hovertemplate='t = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'), 1, c)
    fig.update_xaxes(title_text='t', row=1, col=c)
fig.update_layout(title='Выход сети как сумма сдвинутых и масштабированных гиперболических тангенсов',
                  height=520, legend=dict(orientation='h', x=0.5, xanchor='center', y=-0.18))
save(fig, '10_neuron_contributions', width=1200, height=520)
fig.show()

## 6. Влияние коэффициента обучения lr

Методические указания допускают $\eta \in [0{,}01;\ 0{,}5]$. Для $S = 10$ и бюджета 600 эпох сравним траектории ошибки (медиана по 10 инициализациям).

In [18]:
LR_LIST = [0.01, 0.05, 0.1, 0.2, 0.25, 0.3, 0.5]
lr_rows, lr_curves = [], {}
for lr_ in LR_LIST:
    recs = [train(TwoLayerNet(10, t, y, seed=s), t, y, fcn='traingd', lr=lr_, epochs=600, goal=1e-2) for s in SEEDS]
    fin = np.array([r.final for r in recs])
    diverged = np.array([r.stop == 'divergence' for r in recs])
    lr_rows.append(dict(lr=lr_, **{'MSE медиана': np.median(fin) if diverged.sum() < 5 else np.nan,
                                   'цель достигнута': f'{sum(r.stop == "goal" for r in recs)}/10',
                                   'эпох до цели (медиана)': (int(np.median([r.epochs for r in recs if r.stop == 'goal']))
                                                              if any(r.stop == 'goal' for r in recs) else '—'),
                                   'расходимость': f'{sum(r.stop == "divergence" for r in recs)}/10'}))
    lr_curves[lr_] = recs[int(np.argsort([r.final if np.isfinite(r.final) else 1e99 for r in recs])[5])]
df_lr = pd.DataFrame(lr_rows)
df_lr.style.format({'MSE медиана': '{:.4f}'}).hide(axis='index')

lr,MSE медиана,цель достигнута,эпох до цели (медиана),расходимость
0.010000,0.2342,0/10,—,0/10
0.050000,0.0493,0/10,—,0/10
0.100000,0.0262,2/10,477,0/10
0.200000,0.0100,8/10,332,0/10
0.250000,nan,0/10,—,10/10
0.300000,nan,0/10,—,10/10
0.500000,nan,0/10,—,10/10


In [19]:
fig = go.Figure()
lr_colors = ['#b7d3f6', '#86b6ef', '#5598e7', '#2a78d6', '#1c5cab', '#104281', GOAL_COLOR]  # lr = 0.5 — красный: расходимость
for lr_, col in zip(LR_LIST, lr_colors):
    pr = np.array(lr_curves[lr_].perf)
    pr = pr[np.isfinite(pr)]
    fig.add_trace(go.Scatter(x=np.arange(len(pr)), y=pr, mode='lines', name=f'lr = {lr_}',
                             line=dict(color=col, width=2.5 if lr_ in (0.05, 0.5) else 2,
                                       dash='dot' if lr_ == 0.5 else None),
                             hovertemplate=f'lr = {lr_}<br>' + 'эпоха %{x}<br>MSE = %{y:.4g}<extra></extra>'))
fig.add_hline(y=1e-2, line=dict(color=GOAL_COLOR, dash='dash', width=1),
              annotation_text='goal', annotation_position='bottom left', annotation_font_color=GOAL_COLOR)
fig.update_layout(title='traingd, S = 10: влияние коэффициента обучения на траекторию ошибки',
                  xaxis_title='эпоха', yaxis_title='MSE (log)', yaxis_type='log', yaxis_range=[-2.3, 3.2],
                  hovermode='x unified')
save(fig, '11_learning_rate')
fig.show()

**Почему граница устойчивости проходит около 0,2.** Для квадратичной модели ошибки градиентный спуск сходится, только если $\eta < 2/\lambda_{\max}$, где $\lambda_{\max}$ — наибольшее собственное число гессиана. Оценим гессиан приближением Гаусса — Ньютона $H \approx \tfrac{2}{N} J^\top J$.

In [20]:
lam_rows = []
for s in range(5):
    net = TwoLayerNet(10, t, y, seed=s)
    xn_ = net.sx.apply(t)
    ev0 = np.linalg.eigvalsh(2 / len(t) * net.jacobian(xn_).T @ net.jacobian(xn_))
    train(net, t, y, fcn='traingd', lr=0.05, epochs=600, goal=1e-2)
    ev1 = np.linalg.eigvalsh(2 / len(t) * net.jacobian(xn_).T @ net.jacobian(xn_))
    lam_rows.append(dict(seed=s, **{'λmax (старт)': ev0.max(), 'λmax (600 эп.)': ev1.max(),
                                    '2/λmax': 2 / ev1.max(), 'λmin (600 эп.)': ev1.min(),
                                    'число обусловленности': ev1.max() / max(ev1.min(), 1e-300)}))
df_lam = pd.DataFrame(lam_rows)
LR_CRIT = float(df_lam['2/λmax'].median())
print(f'Оценка критического шага: 2/λmax ≈ {LR_CRIT:.3f}')
df_lam.style.format({'λmax (старт)': '{:.2f}', 'λmax (600 эп.)': '{:.2f}', '2/λmax': '{:.3f}',
                     'λmin (600 эп.)': '{:.1e}', 'число обусловленности': '{:.1e}'}).hide(axis='index')

Оценка критического шага: 2/λmax ≈ 0.190


seed,λmax (старт),λmax (600 эп.),2/λmax,λmin (600 эп.),число обусловленности
0,10.49,10.50,0.190,1.3e-11,8.1e+11
1,10.47,10.51,0.190,1.2e-11,9.1e+11
2,10.47,10.55,0.190,2.2e-11,4.8e+11
3,10.47,10.48,0.191,2.3e-11,4.5e+11
4,10.49,10.53,0.190,8.0e-11,1.3e+11


## 7. Недообучение и переобучение

На точной (незашумлённой) функции с плотной сеткой переобучения почти не видно: 141 точка жёстко фиксирует форму кривой. Чтобы показать эффект наглядно, добавим к обучающим значениям шум $\mathcal N(0,\ 0{,}7^2)$ и будем оценивать качество по **истинной** функции на контрольной сетке. Сети обучаются методом `trainlm` до минимума ошибки на обучающих точках (1000 эпох).

In [21]:
rng_noise = np.random.default_rng(2026)
SIGMA = 0.7
y_noisy = y + rng_noise.normal(0, SIGMA, len(y))
print(f'Дисперсия шума σ² = {SIGMA ** 2:.2f} — ниже этого уровня MSE на обучении означает подгонку под шум')

noise_rows, noise_models = [], {}
for S in S_ALL:
    best = None
    for s in range(5):
        net = TwoLayerNet(S, t, y_noisy, seed=s)
        rec = train(net, t, y_noisy, fcn='trainlm', epochs=1000, goal=0)
        if best is None or rec.final < best[1].final:
            best = (net, rec)
    net, rec = best
    noise_models[S] = net
    noise_rows.append(dict(S=S, train=rec.final, test=mse(net.sim(t_dense), y_dense)))
df_noise = pd.DataFrame(noise_rows)
S_best_noise = int(df_noise.loc[df_noise.test.idxmin(), 'S'])
print(f'Лучшее S по ошибке на истинной функции: S = {S_best_noise}')

# Ранний останов (как dividerand 70/15/15 + max_fail = 6 в MATLAB) для S = 30
perm = np.random.default_rng(7).permutation(len(t))
i_tr, i_val = np.sort(perm[:int(0.7 * len(t))]), np.sort(perm[int(0.7 * len(t)):int(0.85 * len(t))])
es_tests = []
for s in range(5):
    net = TwoLayerNet(30, t[i_tr], y_noisy[i_tr], seed=s)
    rec_es = train(net, t[i_tr], y_noisy[i_tr], fcn='trainlm', epochs=1000, goal=0,
                   val=(t[i_val], y_noisy[i_val]), max_fail=6)
    es_tests.append((mse(net.sim(t_dense), y_dense), net, rec_es))
es_test, net_es, rec_es = sorted(es_tests, key=lambda z: z[0])[2]   # медианный из 5
print(f'S = 30 без регуляризации: MSE(истинная) = {df_noise.set_index("S").test[30]:.3f}')
print(f'S = 30 с ранним остановом: MSE(истинная) = {es_test:.3f} (останов на эпохе {rec_es.epochs}, {rec_es.stop})')
df_noise.iloc[[0, 1, 2, 4, 9, 19, 29]].style.format({'train': '{:.3f}', 'test': '{:.3f}'}).hide(axis='index')

Дисперсия шума σ² = 0.49 — ниже этого уровня MSE на обучении означает подгонку под шум


Лучшее S по ошибке на истинной функции: S = 3
S = 30 без регуляризации: MSE(истинная) = 4.062
S = 30 с ранним остановом: MSE(истинная) = 0.374 (останов на эпохе 10, val_fail)


S,train,test
1,1.835,1.512
2,1.041,0.684
3,0.531,0.060
5,0.487,0.066
10,0.360,0.231
20,0.249,0.350
30,0.170,4.062


In [22]:
fig = make_subplots(rows=1, cols=3, shared_yaxes=True, horizontal_spacing=0.04,
                    subplot_titles=[f'S = 2 — недообучение', f'S = {S_best_noise} — баланс', 'S = 30 — переобучение'])
for c, S in enumerate((2, S_best_noise, 30), start=1):
    fig.add_trace(go.Scatter(x=t, y=y_noisy, mode='markers', name='зашумлённые точки', legendgroup='p',
                             showlegend=c == 1, marker=dict(color=MUTED, size=5, opacity=0.8)), 1, c)
    fig.add_trace(go.Scatter(x=t_dense, y=y_dense, mode='lines', name='истинная f(t)', legendgroup='f',
                             showlegend=c == 1, line=dict(color=INK, width=1.5, dash='dash')), 1, c)
    row = df_noise.set_index('S').loc[S]
    fig.add_trace(go.Scatter(x=t_dense, y=noise_models[S].sim(t_dense), mode='lines', name='выход сети',
                             legendgroup='nn', showlegend=c == 1, line=dict(color=PALETTE[1], width=2.5),
                             hovertemplate=f'S = {S}<br>' + 't = %{x:.2f}<br>ŷ = %{y:.3f}<extra></extra>'), 1, c)
    fig.add_annotation(text=f'обуч.: {row.train:.3f}<br>истин.: {row.test:.3f}', xref=f'x{c if c > 1 else ""} domain',
                       yref='y domain', x=0.03, y=0.97, showarrow=False, align='left', xanchor='left',
                       yanchor='top', font=dict(size=12, color=INK2), bgcolor='rgba(252,252,251,0.85)')
    fig.update_xaxes(title_text='t', row=1, col=c)
fig.update_yaxes(range=[ylo - 1, yhi + 1])
fig.update_layout(title='Данные с шумом σ = 0.7: признаки недообучения и переобучения', height=500,
                  legend=dict(orientation='h', x=0.5, xanchor='center', y=-0.2))
save(fig, '12_under_overfitting', width=1300, height=500)
fig.show()

In [23]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=df_noise.S, y=df_noise.train, mode='lines+markers', name='MSE на обучающих (шумных) точках',
                         line=dict(color=PALETTE[0], width=2), marker=dict(size=7, line=dict(color=SURFACE, width=1.5))))
fig.add_trace(go.Scatter(x=df_noise.S, y=df_noise.test, mode='lines+markers', name='MSE относительно истинной f(t)',
                         line=dict(color=PALETTE[1], width=2), marker=dict(size=7, symbol='diamond',
                                                                            line=dict(color=SURFACE, width=1.5))))
fig.add_trace(go.Scatter(x=[30], y=[es_test], mode='markers+text', name='S = 30 + ранний останов',
                         marker=dict(size=13, symbol='star', color=PALETTE[2], line=dict(color=SURFACE, width=1.5)),
                         text=[f'  {es_test:.3f}'], textposition='middle left', textfont=dict(color=INK2)))
fig.add_hline(y=SIGMA ** 2, line=dict(color=MUTED, dash='dot', width=1.5),
              annotation_text='дисперсия шума σ²', annotation_position='top left', annotation_font_color=INK2)
fig.update_layout(title='Ошибка на обучении падает всегда, ошибка на истинной функции — U-образно',
                  xaxis_title='число нейронов S', yaxis_title='MSE (log)', yaxis_type='log', xaxis=dict(dtick=2),
                  hovermode='x unified', legend=dict(x=0.01, y=0.99))
save(fig, '13_bias_variance')
fig.show()

## 8. Интерактивное приложение

Файл `app/nn_trainer.html` — самостоятельная веб-страница (JavaScript + Plotly.js), которая реализует требования раздела 3 методических указаний:
- ползунок **числа нейронов** 1…30, ручной ввод **lr** $\in [0{,}01;\ 0{,}5]$ и **goal** $\in [10^{-4};\ 10^{-1}]$, а также `epochs`, `show` и выбор алгоритма (`traingd` по умолчанию, `traingdm`, `traingdx`, `trainlm`);
- **мультипликация**: кривая аппроксимации перерисовывается каждые `show` эпох, прошлые положения кривой остаются бледным «следом»;
- **синхронный график ошибки** выстраивается точка за точкой;
- ползунок **«Задержка (мс)»** и кнопки **«Старт»**, **«Пауза»**, **«Сброс»**;
- дополнительно: выбор варианта (1–14), график вкладов нейронов и кнопка «Серия S = 1, 2, 5, 10, 20, 30».

Открыть: двойным щелчком по файлу в браузере (нужен интернет для загрузки Plotly.js с CDN).

In [24]:
from IPython.display import IFrame
IFrame('app/nn_trainer.html', width='100%', height=900)

## 9. Сводка результатов

In [25]:
summary = df_algos.pivot(index='S', columns='алгоритм', values='MSE медиана')[
    ['traingd', 'traingd (≤20000 эп.)', 'traingdx', 'trainlm']]
summary['предел (trainlm, 500 эп.)'] = df_cap.set_index('S').train.reindex(summary.index)
fig = go.Figure(go.Table(
    header=dict(values=['S'] + list(summary.columns), fill_color='#eef3fb', line_color=GRID,
                font=dict(color=INK, size=13), align='center', height=34),
    cells=dict(values=[summary.index] + [[f'{v:.4f}' if v >= 1e-3 else f'{v:.1e}' for v in summary[c]]
                                         for c in summary.columns],
               fill_color=SURFACE, line_color=GRID, font=dict(color=INK2, size=13), align='center', height=30)))
fig.update_layout(title='Медианная MSE на обучающей сетке (10 инициализаций)', height=380, margin=dict(t=60, b=10))
save(fig, '14_results_table', png=False)
fig.show()

# экспорт таблиц в отчёт
def fmt(v, prec=4):
    return f'{v:.{prec}f}' if v >= 10 ** (-prec + 1) else f'{v:.1e}'.replace('e-0', 'e-')


t1 = df_gd.copy()
t1 = pd.DataFrame({'$S$': t1.S, 'Параметров': t1['параметров'],
                   'MSE медиана': t1['MSE медиана'].map(fmt), 'MSE лучшая': t1['MSE лучшая'].map(fmt),
                   'MSE худшая': t1['MSE худшая'].map(fmt), 'MSE контроль': t1['MSE контроль'].map(fmt),
                   '$R^2$': t1['R² контроль'].map(lambda v: f'{v:.4f}'), 'Цель': t1['цель достигнута']})
to_latex(t1, 'tab_series_gd', 'Серия экспериментов по заданию (traingd, $\\eta=0{,}05$, 600 эпох, 10 инициализаций)',
         'tab:series')

t2 = df_algos[df_algos['алгоритм'] != 'traingd'].copy()
t2 = pd.DataFrame({'Алгоритм': t2['алгоритм'].str.replace('≤', '$\\le$', regex=False), '$S$': t2.S,
                   'MSE медиана': t2['MSE медиана'].map(fmt), 'MSE контроль': t2['MSE контроль'].map(fmt),
                   'Эпох': t2['эпох (медиана)'], 'Цель': t2['цель достигнута']})
to_latex(t2, 'tab_algorithms', 'Доведение обучения до целевой ошибки $10^{-2}$ (медиана по 10 инициализациям)',
         'tab:algos', col_format='lccccc')

t3 = df_cap[df_cap.S.isin([1, 2, 3, 4, 5, 10, 20, 30])]
t3 = pd.DataFrame({'$S$': t3.S, 'MSE обучение': t3.train.map(lambda v: f'{v:.1e}'),
                   'MSE контроль': t3.test.map(lambda v: f'{v:.1e}'), 'max$|e|$': t3.maxerr.map(lambda v: f'{v:.1e}')})
to_latex(t3, 'tab_capacity', 'Предельная точность сети с $S$ нейронами (trainlm, лучший из 10 запусков)', 'tab:capacity')

t4 = pd.DataFrame({'$\\eta$': df_lr.lr.map(lambda v: f'{v:g}'), 'MSE медиана': df_lr['MSE медиана'].map(lambda v: '---' if np.isnan(v) else fmt(v)),
                   'Цель достигнута': df_lr['цель достигнута'], 'Эпох до цели': df_lr['эпох до цели (медиана)'],
                   'Расходимость': df_lr['расходимость']})
to_latex(t4, 'tab_lr', 'Влияние коэффициента обучения (traingd, $S=10$, 600 эпох, 10 инициализаций)', 'tab:lr')

t5 = df_noise[df_noise.S.isin([1, 2, 3, 5, 8, 10, 15, 20, 30])]
t5 = pd.DataFrame({'$S$': t5.S.astype(str), 'MSE на шумных точках': t5.train.map(lambda v: f'{v:.3f}'),
                   'MSE относительно $f(t)$': t5.test.map(lambda v: f'{v:.3f}')})
t5 = pd.concat([t5, pd.DataFrame([{'$S$': '30 + ранний останов', 'MSE на шумных точках': f'{rec_es.perf[-1]:.3f}',
                                    'MSE относительно $f(t)$': f'{es_test:.3f}'}])], ignore_index=True)
to_latex(t5, 'tab_noise', 'Недообучение и переобучение на зашумлённых данных ($\\sigma = 0{,}7$, trainlm)', 'tab:noise')
print('Таблицы для отчёта:', sorted(p.name for p in TABLES.glob('*.tex')))
summary.style.format('{:.2e}')

Таблицы для отчёта: ['tab_algorithms.tex', 'tab_capacity.tex', 'tab_lr.tex', 'tab_noise.tex', 'tab_series_gd.tex']


алгоритм,traingd,traingd (≤20000 эп.),traingdx,trainlm,"предел (trainlm, 500 эп.)"
S,,,,,
1,4.08e+00,1.55e+00,1.51e+00,1.50e+00,1.50e+00
2,2.18e+00,8.65e-01,7.76e-01,6.75e-01,6.75e-01
5,4.84e-02,1.00e-02,9.96e-03,7.05e-03,7.27e-09
10,4.93e-02,1.00e-02,9.91e-03,2.99e-03,2.45e-09
20,6.70e-02,1.00e-02,9.93e-03,2.73e-03,9.67e-10
30,9.58e-02,1.00e-02,9.97e-03,9.10e-04,3.37e-10


## Выводы

1. **Протокол по заданию** (`traingd`, $\eta = 0{,}05$, 600 эпох, `goal = 1e-2`) даёт хорошую, но не доведённую до цели аппроксимацию: при $S = 10$ MSE $= 0{,}045$, $R^2 = 0{,}995$, максимальное отклонение $0{,}44$ при размахе функции $12{,}4$. Цель $10^{-2}$ за 600 эпох не достигнута ни в одном из 60 запусков серии — ограничивает не архитектура, а бюджет обучения.

2. **Влияние числа нейронов.** $S = 1$ (MSE $\approx 4{,}1$) и $S = 2$ ($\approx 2{,}2$) — явное недообучение: одна-две сигмоиды не могут повторить три экстремума функции, и это предел ёмкости, а не оптимизации (даже `trainlm` до сходимости даёт лишь $1{,}50$ и $0{,}68$). Начиная с $S = 3$ достижима MSE $\approx 10^{-4}$, с $S = 5$ — $\approx 10^{-8}$: для гладкой функции с тремя экстремумами достаточно 3–5 нейронов.

3. **Больше нейронов — не лучше при фиксированном бюджете.** В протоколе по заданию медианная MSE растёт с $0{,}048$ ($S = 5$) до $0{,}096$ ($S = 30$), на графиках видна мелкая «рябь». Это недоученность, а не переобучение: ошибка на контрольной сетке равна ошибке на обучающей. Инициализация Нгуена — Уидроу даёт крутые сигмоиды ($|w_1| = 0{,}7S$), и 600 малых шагов не успевают их согласовать.

4. **Нормальное обучение достигает цели при любом $S \ge 5$** (10/10 запусков): `traingd` — за 2360–3960 эпох, `traingdx` — за 117–386, `trainlm` — за 1–5 эпох. Причина медленности `traingd` — сильно вытянутая («овражная») поверхность ошибки: число обусловленности гессиана $\sim 10^{11}$–$10^{12}$.

5. **Коэффициент обучения.** Рост $\eta$ от $0{,}01$ до $0{,}2$ ускоряет сходимость (MSE за 600 эпох: $0{,}23 \to 0{,}01$), но при $\eta \ge 0{,}25$ обучение расходится во всех запусках. Граница совпадает с теоретической оценкой $\eta < 2/\lambda_{\max} \approx 0{,}19$.

6. **Переобучение** проявляется на зашумлённых данных: при $S = 30$ MSE на обучающих точках ($0{,}17$) ниже дисперсии шума ($0{,}49$), т.е. сеть подгоняется под шум, а ошибка относительно истинной функции вырастает до $4{,}06$ (выбросы кривой между точками). Лучший баланс — $S = 3$ (MSE $0{,}06$ относительно истинной функции). Ранний останов по валидационной выборке снижает ошибку сети с $S = 30$ в 11 раз ($4{,}06 \to 0{,}37$).

7. Результаты подтверждают теорему об универсальной аппроксимации: одного скрытого слоя хватает для приближения функции с любой точностью, но практическая точность определяется балансом числа нейронов, алгоритма обучения и объёма/качества данных.

## Ответы на контрольные вопросы

**1. Какова математическая и алгоритмическая роль нелинейной функции активации в скрытом слое при решении задач аппроксимации?**

Без нелинейности сеть вырождается в линейную: $W_2(W_1 t + b_1) + b_2 = W t + b$ — композиция линейных отображений линейна, и любое число нейронов даёт лишь прямую. Нелинейная `tansig` превращает каждый нейрон в гладкую «ступеньку» $\tanh(w t + b)$ с настраиваемыми положением ($-b/w$), крутизной ($w$) и высотой ($w_2$). Выходной линейный слой складывает эти ступеньки, и из их суммы можно собрать кривую любой формы (график вкладов нейронов в разделе 5). Алгоритмически важно, что `tansig` дифференцируема, а её производная просто выражается через значение: $\tanh'(z) = 1 - \tanh^2(z)$ — это делает обратное распространение дешёвым. Ограниченность и симметричность относительно нуля улучшают обусловленность задачи по сравнению с логистической сигмоидой.

**2. Опишите физический и математический смысл алгоритма обратного распространения ошибки (traingd). Как коэффициент обучения (lr) влияет на сходимость траектории?**

Математически `traingd` — градиентный спуск $\theta \leftarrow \theta - \eta\,\nabla E(\theta)$ по функции ошибки MSE. Обратное распространение — эффективный способ вычислить градиент по правилу дифференцирования сложной функции: ошибка выхода $\delta_2 = \tfrac{2}{N}(\hat y - y)$ передаётся назад через веса $W_2$ и умножается на производную активации, $\delta_1 = (\delta_2 W_2) \odot (1 - a_1^2)$; градиент по весу равен произведению «сигнала ошибки» на вход этого веса. Физически — это движение шарика по поверхности ошибки в направлении наискорейшего спуска, где каждый вес получает поправку, пропорциональную своему вкладу в ошибку.

`lr` — длина шага. Малый шаг (0,01) даёт устойчивую, но очень медленную траекторию (MSE $0{,}23$ за 600 эпох). Увеличение шага ускоряет спуск ($\eta = 0{,}2$ — цель за ~330 эпох), но если шаг больше $2/\lambda_{\max}$ (здесь $\approx 0{,}19$), траектория «перепрыгивает» дно оврага, колеблется с растущей амплитудой и расходится ($\eta \ge 0{,}25$). Поверхность ошибки сильно вытянута (число обусловленности $\sim 10^{11}$), поэтому один постоянный шаг не подходит одновременно для крутых и пологих направлений — отсюда медленность `traingd` и выигрыш `traingdx` (адаптивный шаг и момент) и `trainlm` (учёт кривизны).

**3. Сформулируйте суть теоремы об универсальной аппроксимации (теорема Цыбенко / Колмогорова). Что она накладывает на количество слоёв сети?**

Теорема Цыбенко (1989), обобщённая Хорником (1991): для любой непрерывной функции $f$ на компакте $K \subset \mathbb R^n$ и любого $\varepsilon > 0$ найдётся сеть с **одним** скрытым слоем из конечного числа $S$ нейронов с сигмоидальной активацией и линейным выходом, такая что $\max_{x \in K} |f(x) - \hat f(x)| < \varepsilon$. Теорема Колмогорова — Арнольда утверждает, что любую непрерывную функцию $n$ переменных можно точно представить суперпозицией непрерывных функций одной переменной и сложения ($2n+1$ внутренних функций). Следствие для архитектуры: двух слоёв (скрытый нелинейный + линейный выход) **достаточно** — дополнительные слои не нужны для самой возможности аппроксимации. Однако теорема не говорит, сколько нейронов потребуется, не гарантирует, что их найдёт градиентный спуск, и ничего не говорит о поведении вне обучающих точек. Эксперимент это иллюстрирует: при $S \ge 3$ точность $10^{-2}$ достижима, но `traingd` за 600 эпох её не находит.

**4. Каковы явные визуальные признаки недообучения и переобучения (overfitting) сети, наблюдаемые на графике аппроксимации при манипуляциях с размером скрытого слоя?**

*Недообучение* ($S = 1, 2$ или слишком короткое обучение): кривая сети слишком «жёсткая» — гладкая, с меньшим числом изгибов, чем у функции; пропущены экстремумы (у $S = 1$ — почти наклонная прямая, у $S = 2$ — две ступеньки с плато между ними); ошибка систематическая, большая и одинаковая на обучающих и контрольных точках.

*Переобучение* ($S = 30$ на зашумлённых данных): кривая проходит почти через каждую обучающую точку, повторяя шум; между точками появляются осцилляции и резкие выбросы (у нас — узкий выброс от $-23$ до $+32$ около $t \approx 0{,}1$, между соседними обучающими точками, при размахе функции около 12); ошибка на обучающих точках ниже уровня шума, а относительно истинной функции — многократно выше. Признак — большой разрыв между ошибкой на обучающей и контрольной выборке.

Важно отличать от переобучения мелкую рябь, которая видна у $S = 20, 30$ после 600 эпох `traingd`: здесь ошибка на контрольной сетке равна ошибке на обучающей, это недоученность, и она исчезает при продолжении обучения.

**5. Объясните назначение параметров `net.trainParam.show` и `net.trainParam.goal`.**

`show = 50` — период вывода промежуточных результатов: каждые 50 эпох печатается строка протокола (эпоха, MSE, градиент) и обновляется график обучения; на сам процесс обучения не влияет. В нашей реализации по нему делается снимок кривой для мультипликации. `goal = 1e-2` — целевое значение функции качества (MSE): обучение прекращается досрочно, как только $E \le$ `goal`. Это критерий «достаточной точности», экономящий эпохи; наряду с ним работают лимит `epochs` и порог `min_grad`. Слишком малый `goal` на шумных данных толкает сеть к переобучению: ошибку ниже уровня шума можно получить, только подгоняясь под шум.